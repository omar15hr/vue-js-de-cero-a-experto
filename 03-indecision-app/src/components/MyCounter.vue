<template>
  <section>
    <h3>Counter: {{ counter }}</h3>
    <h3 data-testid="square-label">Square: {{ squareCounter }}</h3>

    <div>
      <button @click="counter++">+1</button>
      <button @click="counter--">-1</button>
    </div>
  </section>
</template>

<script lang="ts" setup>
// import { computed, ref } from 'vue';
import { useCounter } from '../composables/useCounter';

interface Props {
  value?: number;
}
const props = defineProps<Props>();

const { counter, squareCounter } = useCounter(props.value);


// const counter = ref(props.value ?? 5);
// const squareCounter = computed(() => counter.value * counter.value);
</script>