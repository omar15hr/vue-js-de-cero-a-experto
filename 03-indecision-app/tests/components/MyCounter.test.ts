import { describe, expect, test } from 'vitest';
import { mount } from '@vue/test-utils';
import MyCounter from '@/components/MyCounter.vue';

describe('MyCounter component', () => {
  const value = 5;

  const wrapper = mount(MyCounter, {
    props: {
      value: value,
    },
  });

  test('should match snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  test('renders the counter value correctly', () => {
    const [counterLabel, squareLabel] = wrapper.findAll('h3');

    expect(wrapper.find('h3').text()).toContain(`Counter: ${value}`);
    expect(wrapper.find('[data-testid="square-label"]').text()).toContain(
      `Square: ${value * value}`,
    );
    expect(counterLabel.text()).toContain(`Counter: ${value}`);
    expect(squareLabel.text()).toContain(`Square: ${value * value}`);
  });

  test('increments the counter value when the button is clicked', async() => {
    const [counterLabel, squareLabel] = wrapper.findAll('h3');

    const btnIncrement = wrapper.find('button');
    await btnIncrement.trigger('click');
    
    expect(counterLabel.text()).toContain(`Counter: ${value + 1}`);
    expect(squareLabel.text()).toContain(`Square: ${(value+1) * (value+1)}`);
  });

  test('decrements the counter when -1 button is clicked twice', async () => {
    const value = 5;
    const wrapper = mount(MyCounter, {
      props: {
        value: value,
      },
    });

    const [counterLabel, squareLabel] = wrapper.findAll('h3');

    const [, btnDecrement] = wrapper.findAll('button');
    await btnDecrement.trigger('click');
    await btnDecrement.trigger('click');

    expect(counterLabel.text()).toContain(`Counter: ${value - 2}`);
    expect(squareLabel.text()).toContain(`Square: ${(value - 2) * (value - 2)}`);
  });
});
