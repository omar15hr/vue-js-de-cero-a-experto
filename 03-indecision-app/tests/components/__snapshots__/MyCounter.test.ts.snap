// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MyCounter component > should match snapshot 1`] = `
"<section>
  <h3>Counter: 5</h3>
  <h3 data-testid="square-label">Square: 25</h3>
  <div><button>+1</button><button>-1</button></div>
</section>"
`;
